File: src/main/scala/stirling/fix/messages/fix44/burgundy/ExecutionReport.scala
/*
 * Copyright 2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package stirling.fix.messages.fix44.burgundy

import stirling.fix.messages.{
  AbstractMessage,
  ExecutionReport => ExecutionReportTrait,
  MessageHeader,
  MessageVisitor,
  RepeatingGroup,
  RepeatingGroupInstance,
  Required
}
import stirling.fix.tags.fix42.{
  Account,
  AvgPx,
  ClOrdID,
  CumQty,
  Currency,
  ExecID,
  ExecInst,
  ExecRefID,
  ExecType,
  ExpireTime,
  HandlInst,
  LastPx,
  LastShares,
  LeavesQty,
  MaxFloor,
  MaxShow,
  MinQty,
  OrdRejReason,
  OrdStatus,
  OrdType,
  OrderID,
  OrderQty,
  OrigClOrdID,
  Price,
  SecondaryOrderID,
  SecurityID,
  Side,
  Symbol,
  Text,
  TimeInForce,
  TradeDate,
  TransactTime
}
import stirling.fix.tags.fix43.{
  AccountType,
  ExecRestatementReason,
  MassStatusReqID,
  NoPartyIDs,
  OrderCapacity,
  PartyID,
  PartyIDSource,
  PegOffsetValue,
  SecondaryExecID,
  SecurityIDSource
}
import stirling.fix.tags.fix44.{
  LastLiquidityInd,
  LastRptRequested,
  PartyRole,
  PegMoveType,
  PegOffsetType,
  PegScope,
  PeggedPrice,
  TotNumReports
}
import stirling.fix.tags.fix44.burgundy.{
  OrderRestrictions,
  TrdType
}
import stirling.fix.tags.fix50.{
  DisplayMethod,
  MatchType,
  DisplayLowQty,
  DisplayHighQty,
  DisplayMinIncr
}

class ExecutionReport(header: MessageHeader) extends AbstractMessage(header) with ExecutionReportTrait with Groups {
  field(Account.Tag, Required.NO)
  field(AvgPx.Tag)
  field(ClOrdID.Tag, Required.NO)
  field(CumQty.Tag)
  field(Currency.Tag, Required.NO)
  field(ExecID.Tag)
  field(ExecInst.Tag, Required.NO)
  field(ExecRefID.Tag, Required.NO)
  field(HandlInst.Tag, Required.NO)
  field(LastPx.Tag, Required.NO)
  field(LastShares.Tag, Required.NO)
  field(OrderID.Tag)
  field(OrderQty.Tag)
  field(OrdStatus.Tag)
  field(OrdType.Tag, Required.NO)
  field(OrigClOrdID.Tag, Required.NO)
  field(Price.Tag, Required.NO)
  field(Side.Tag, Required.NO)
  field(Symbol.Tag)
  field(SecurityIDSource.Tag, Required.NO)
  field(SecurityID.Tag, Required.NO)
  field(Text.Tag, Required.NO)
  field(TimeInForce.Tag, Required.NO)
  field(TransactTime.Tag, Required.NO)
  field(TradeDate.Tag, Required.NO)
  field(OrdRejReason.Tag, Required.NO)
  field(MinQty.Tag, Required.NO)
  field(DisplayMethod.Tag, Required.NO)
  val requiredWhenDisplayMethodIsRandom = new Required() {
    override def isRequired(): Boolean = getEnum(DisplayMethod.Tag) == DisplayMethod.Random
  }
  field(DisplayLowQty.Tag, requiredWhenDisplayMethodIsRandom)
  field(DisplayHighQty.Tag, requiredWhenDisplayMethodIsRandom)
  field(DisplayMinIncr.Tag, Required.NO)
  field(MaxFloor.Tag, Required.NO)
  field(ExpireTime.Tag, new Required() {
    override def isRequired(): Boolean = getEnum(TimeInForce.Tag) == TimeInForce.GoodTillDate
  })
  field(ExecType.Tag)
  field(LeavesQty.Tag)
  field(SecondaryOrderID.Tag, Required.NO)
  field(MaxShow.Tag, Required.NO)
  field(PegOffsetValue.Tag, Required.YES)
  field(PegMoveType.Tag, Required.NO)
  field(PegOffsetType.Tag, Required.NO)
  field(PegScope.Tag, Required.NO)
  field(ExecRestatementReason.Tag, Required.NO)
  field(SecondaryExecID.Tag, Required.NO)
  field(OrderCapacity.Tag, Required.NO)
  field(OrderRestrictions.Tag, Required.NO)
  field(MatchType.Tag, Required.NO)
  field(AccountType.Tag, Required.NO)
  field(MassStatusReqID.Tag, Required.NO)
  parties(Required.NO)
  field(TrdType.Tag, new Required() {
    override def isRequired(): Boolean = getEnum(MatchType.Tag) == MatchType.TwoPartyTradeReport
  })
  field(PeggedPrice.Tag, Required.NO)
  field(LastLiquidityInd.Tag, Required.NO)
  field(TotNumReports.Tag, Required.NO)
  field(LastRptRequested.Tag, Required.NO)

  override def apply(visitor: MessageVisitor) = visitor.visit(this)
}

File: core/src/main/java/stirling/fix/messages/fix42/chix/europe/ExecutionReport.java
/*
 * Copyright 2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package stirling.fix.messages.fix42.chix.europe;

import stirling.fix.messages.MessageHeader;
import stirling.fix.messages.MessageVisitor;
import stirling.fix.messages.Required;

import stirling.fix.tags.fix42.Account;
import stirling.fix.tags.fix42.AvgPx;
import stirling.fix.tags.fix42.ClOrdID;
import stirling.fix.tags.fix42.ClientID;
import stirling.fix.tags.fix42.ContraBroker;
import stirling.fix.tags.fix42.CumQty;
import stirling.fix.tags.fix42.Currency;
import stirling.fix.tags.fix42.ExecID;
import stirling.fix.tags.fix42.ExecRefID;
import stirling.fix.tags.fix42.ExecTransType;
import stirling.fix.tags.fix42.ExecType;
import stirling.fix.tags.fix42.LastPx;
import stirling.fix.tags.fix42.LastShares;
import stirling.fix.tags.fix42.LeavesQty;
import stirling.fix.tags.fix42.MaxFloor;
import stirling.fix.tags.fix42.MinQty;
import stirling.fix.tags.fix42.NoContraBrokers;
import stirling.fix.tags.fix42.OrdStatus;
import stirling.fix.tags.fix42.OrdType;
import stirling.fix.tags.fix42.OrderCapacity;
import stirling.fix.tags.fix42.OrderID;
import stirling.fix.tags.fix42.OrderQty;
import stirling.fix.tags.fix42.OrigClOrdID;
import stirling.fix.tags.fix42.Price;
import stirling.fix.tags.fix42.SecurityID;
import stirling.fix.tags.fix42.Symbol;
import stirling.fix.tags.fix42.Text;
import stirling.fix.tags.fix42.TimeInForce;
import stirling.fix.tags.fix42.TradeDate;
import stirling.fix.tags.fix42.TransactTime;
import stirling.fix.tags.fix42.chix.europe.IDSource;
import stirling.fix.tags.fix42.chix.europe.TradeLiquidityIndicator;
import stirling.fix.tags.fix42.LastCapacity;
import stirling.fix.tags.fix42.Side;

public class ExecutionReport extends stirling.fix.messages.AbstractMessage implements stirling.fix.messages.ExecutionReport {
    public ExecutionReport(MessageHeader header) {
        super(header);

        field(Account.Tag(), Required.NO);
        field(AvgPx.Tag());
        field(ClOrdID.Tag(), Required.NO);
        field(CumQty.Tag());
        field(Currency.Tag(), Required.NO);
        field(ExecID.Tag());
        field(ExecRefID.Tag(), Required.NO);
        field(ExecTransType.Tag());
        field(IDSource.Tag(), Required.NO);
        field(LastCapacity.Tag(), Required.NO);
        field(LastPx.Tag());
        field(LastShares.Tag());
        field(OrderID.Tag());
        field(OrderQty.Tag());
        field(OrdStatus.Tag());
        field(OrdType.Tag(), Required.NO);
        field(OrigClOrdID.Tag(), Required.NO);
        field(Price.Tag(), Required.NO);
        field(OrderCapacity.Tag(), Required.NO);
        field(SecurityID.Tag(), new Required() {
            @Override public boolean isRequired() {
                return hasValue(IDSource.Tag());
            }
        });
        field(Side.Tag());
        field(Symbol.Tag());
        field(Text.Tag(), Required.NO);
        field(TimeInForce.Tag(), Required.NO);
        field(TransactTime.Tag());
        field(TradeDate.Tag(), Required.NO);
        field(ClientID.Tag(), Required.NO);
        field(MinQty.Tag(), Required.NO);
        field(MaxFloor.Tag(), Required.NO);
        field(ExecType.Tag());
        field(LeavesQty.Tag());
        field(NoContraBrokers.Tag(), Required.NO);
        field(ContraBroker.Tag(), Required.NO);
        field(TradeLiquidityIndicator.Tag(), Required.NO);
    }

    @Override public void apply(MessageVisitor visitor) {
        visitor.visit(this);
    }
}
